Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,894 @@
+> Haskell gets some resistance due to the complexity of using monads to control side effects.
+> Wadler tries to appease critics by explaining that "a monad is a monoid in the category of endofunctors, what's the problem?"
+>
+> -- James Iry, [A Brief, Incomplete, and Mostly Wrong History of Programming Languages][brief-incomplete-history]
+
+[brief-incomplete-history]: http://james-iry.blogspot.com/2009/05/brief-incomplete-and-mostly-wrong.html
+
+Before we see how this idea applies to monads, it might be helpful to compare monads to monoids first.
+
+Let us first review the definition of the `Monad` type class.
+In Haskell, it is defined as[^applicative]:
+
+[^applicative]: If you're more familiar with modern Haskell, you might object that actually, the superclass of `Monad` is `Applicative`, but we will gloss over that for the sake of simplicity.
+
+```haskell
+class Functor m => Monad m where
+  return :: a -> m a
+  (>>=) :: m a -> (a -> m b) -> m b
+```
+
+This definition needs to additionally satisfy the monad laws:
+
+```haskell
+(m >>= g) >>= h = m >>= (\x -> g x >>= h) -- associativity
+return a >>= h = h a -- left identity
+m >>= return = m -- right identity
+```
+
+If you look at the monad laws and squint a bit (or perhaps a lot), you might notice that these laws seem somewhat similar to the monoid laws.
+The names of the laws are the same, at least.
+Our choice of the definition makes it a little harder to see the connection, though.
+If we instead rewrite these laws using the Kleisli composition
+
+```haskell
+(>=>) :: Monad m => (a -> m b) -> (b -> m c) -> (a -> m c)
+f >=> g = \x -> f x >>= g
+```
+
+it becomes much more apparent[^equivalence]:
+
+[^equivalence]: The proof of equivalence is left as an exercise to the reader.
+    Here, by equivalence we mean that the monad laws expressed via Kleisli arrows imply the monad laws expressed via `>>=` and vice versa.
+    The proof itself is a straightforward application of equational reasoning.
+    It's easier to start from the Kleisli arrow form.
+
+```haskell
+(f >=> g) >=> h = f >=> (g >=> h) -- associativity
+return >=> f = f -- left identity
+f >=> return = f -- right identity
+```
+
+That is why a monad is indeed a monoid: it satisfies the same laws, and the "values" are of the type `Functor f => a -> f b`.
+Note, though, that this is not the category of endofunctors.
+However, there exists a natural transformation to the category of endofunctors[^endofunctors].
+
+[^endofunctors]: One could alternatively define monads via `join :: m (m a) -> m a` instead of `>>=`, which would make it a little more evident that monads are monoids over endofunctors, but to explain this properly, we would need to use a bit more of category theory jargon.
+    There is an excellent [r/math post][endofunctors] on the subject.
+
+    [endofunctors]: https://www.reddit.com/r/math/comments/ap25mr/a_monad_is_a_monoid_in_the_category_of/
```

Comment:
For some reason gets rendered like this 
<img width="518" alt="image" src="https://user-images.githubusercontent.com/35335196/150527597-60dc3c8b-f376-448d-8d23-b2aa1d8e801a.png">

```suggestion
[endofunctors]: (https://www.reddit.com/r/math/comments/ap25mr/a_monad_is_a_monoid_in_the_category_of/)
```

Reply:
Or just use a regular hyperlink.

Reply:
On that note, what engine are we using? I've assumed pandoc-compatible Markdown, but apparently not. Is there a straightforward way to get a preview locally?

Reply:
I don't know which engine exactly we use. I think @kirelagin would be able to answer this? 

There's probably not a straightforward way to preview locally at this point of time. 🤔 All I can suggest is to use the admin panel of website's staging to submit the post and see what happens. 😅

Reply:
I’m let to the party but we use some generic javascript markdown renderer (not that there is some standard anyway).

Reply:
The problem in the highlighted code fragment is that there is something weird with the indentation? I don’t think reference links like `[...]:` can be indented like that.

Reply:
In case you are really wondering, it’s this one: https://github.com/markdown-it/markdown-it

Reply:
> I don’t think reference links like `[...]:` can be indented like that.

Oh, they absolutely can, at least in some flavours -- because it's in the footnote -- anything indented 4 spaces after the footnote is part of the footnote, and at least Pandoc's footnotes allow arbitrary block content. FWIW this code snippet works as expected with markdown-it-footnote, too.